File: LeetCode/Problems/21.py
# NeetCode - Linked List
# 21. Merge Two Sorted Lists
# Difficulty : Easy
# Algorithm : Linked List
# Time complexity : O(N+M), Space complexity : O(N+M)
# Runtime : 57 ms (31.76%), Memory : 16.45 MB (14.92%)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        # Time complexity : O(N+M), Space complexity : O(N+M)
        def rec(self, list1, list2):
            tmp = ListNode()

            if list1 and list2:
                if list1.val < list2.val:
                    tmp.val = list1.val
                    list1 = list1.next
                else:
                    tmp.val = list2.val
                    list2 = list2.next
            elif list1:
                tmp.val = list1.val
                list1 = list1.next
            elif list2:
                tmp.val = list2.val
                list2 = list2.next
            else:
                return

            node = tmp
            node.next = rec(self, list1, list2)
            return node

        return rec(self, list1, list2)

File: LeetCode/Problems/21.java
/*
21. Merge Two Sorted Lists
Difficulty : Easy
Algorithm : Linked List
Time complexity : O(N), Space complexity : O(1)
Runtime : 0 ms (100%), Memory : 42.22 MB (75.56%)
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if(list1 == null && list2 == null) {return null;}

        ListNode list = null;
        ListNode cur = null;

        while(list1 != null && list2 != null) {
            if(list1.val < list2.val) {
                if(list == null) {
                    list = new ListNode(list1.val);
                    cur = list;
                }
                else {
                    cur.next = new ListNode(list1.val);
                    cur = cur.next;
                }
                list1 = list1.next;
            }
            else {
                if(list == null) {
                    list = new ListNode(list2.val);
                    cur = list;
                }
                else {
                    cur.next = new ListNode(list2.val);
                    cur = cur.next;
                }
                list2 = list2.next;
            }
        }

        if(list1 != null) {
            if(list == null) {list = list1;}
            else {cur.next = list1;}
        }
        else if(list2 != null) {
            if(list == null) {list = list2;}
            else {cur.next = list2;}
        }

        return list;
    }
}
